test/cluster/TypeDBClusterRunner.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import static com.vaticle.typedb.common.collection.Collections.map;
+import static com.vaticle.typedb.common.collection.Collections.pair;
+
+public class TypeDBClusterRunner implements TypeDBRunner {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TypeDBClusterRunner.class);
+
+    protected final Map<Addresses, Map<String, String>> serverOptionsMap;
+    private final TypeDBClusterServerRunner.Factory serverRunnerFactory;
+    protected final Map<Addresses, TypeDBClusterServerRunner> serverRunners;
+
+    public static TypeDBClusterRunner create(Path clusterRunnerDir, int serverCount) {
+        return create(clusterRunnerDir, serverCount, new TypeDBClusterServerRunner.Factory());
+    }
+
+    public static TypeDBClusterRunner create(Path clusterRunnerDir, int serverCount, TypeDBClusterServerRunner.Factory serverRunnerFactory) {
+        Set<Addresses> serverAddressesSet = allocateAddressesSet(serverCount);
+        Map<Addresses, Map<String, String>> serverOptionsMap = new HashMap<>();
+        for (Addresses addrs: serverAddressesSet) {
+            Map<String, String> options = new HashMap<>();
+            options.putAll(ClusterServerOpts.address(addrs));
+            options.putAll(ClusterServerOpts.peers(serverAddressesSet));
+            Path srvRunnerDir = clusterRunnerDir.resolve(addrs.externalString()).toAbsolutePath();
+            options.putAll(
+                    map(
+                            pair(ClusterServerOpts.STORAGE_DATA, srvRunnerDir.resolve("server/data").toAbsolutePath().toString()),
+                            pair(ClusterServerOpts.STORAGE_REPLICATION, srvRunnerDir.resolve("server/replication").toAbsolutePath().toString()),
+                            pair(ClusterServerOpts.STORAGE_USER, srvRunnerDir.resolve("server/user").toAbsolutePath().toString()),
+                            pair(ClusterServerOpts.LOG_OUTPUT_FILE_DIRECTORY, srvRunnerDir.resolve("server/logs").toAbsolutePath().toString())
+                    )
+            );
+            serverOptionsMap.put(addrs, options);
+        }
+        return new TypeDBClusterRunner(serverOptionsMap, serverRunnerFactory);
+    }
+
+    private static Set<Addresses> allocateAddressesSet(int serverCount) {
+        Set<Addresses> addresses = new HashSet<>();
+        for (int i = 0; i < serverCount; i++) {
+            String host = "127.0.0.1";
+            int externalPort = 40000 + i * 1111;
+            int internalPortZMQ = 50000 + i * 1111;
+            int internalPortGRPC = 60000 + i * 1111;
+            addresses.add(Addresses.create(host, externalPort, host, internalPortZMQ, host, internalPortGRPC));
+        }
+        return addresses;
+    }
+
+    private TypeDBClusterRunner(Map<Addresses, Map<String, String>> serverOptionsMap, TypeDBClusterServerRunner.Factory serverRunnerFactory) {
```

Comment:
interesting, so the `Addresses` are now the canonical name/ID for each cluster server?

Reply:
Yes. Each server will be uniquely identified by their address.